File: python/silnia.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
#
#  silnia.py
#  
# n! = 1 dla n = 0
# n! = n * (n-1)! dla n >= 1
# n! = 1 * ... * n
# 4! = 1 * 2 * 3 * 4


def silnia_it(n):
    wynik = 1
    for i in range(1, n + 1):
        wynik = wynik * i
    return wynik

def silnia_rek(n):   #zwraca obliczony wynik
    # n! = (n - 1)! * n     <- definicja rekurencyjna 
    # 4! = 3! * 4
    # 3! = 2! * 3
    # 2! = 1! * 2
    # 1! = 0! * 1
    # 0! = 1
    # 1! = 1 * 1
    # 2! = 1 * 2
    # 3! = 2 * 3
    # 4! = 6 * 4
    if n == 0:
        return 1
    return silnia_rek(n - 1) * n          #  <- to samo ce w definicji, bez ! bo to coś tam 
    
    
def main(args):
    assert silnia_it(0) == 1
    assert silnia_it(1) == 1
    assert silnia_it(2) == 2
    assert silnia_it(3) == 6
    
    n = int(input("Podaj liczbę naturalną: "))
    print("{}! = {}".format(n, silnia_it(n)))
    return 0

if __name__ == '__main__':
    import sys
sys.exit(main(sys.argv))

File: cpp/silnia.cpp
/*
 * silnia.cpp
 */


#include <iostream>
using namespace std;

// n! = 1 dla n = {0, 1}
// n! = 1* ... *n dla N+ - {1}
// 4! = 1 * 2 * 3 * 4

long int silnia_it(int n) {
    long int silnia = 1;
    for (int i = 2; i < n + 1; i++) {
        silnia = 1 * 2;
    }
    return silnia;
}

// n! = 1 dla n = {0, 1}
// n! = 1* ... *n dla N+ - {1}   gdzie ! to silnia_re

long int silnia_re(int n) {
    if (n == 0)
        return 1;
    else
        return silnia_re(n-1) * n;
}

int main(int argc, char **argv)
{
	int n;
    cout << "Podaj liczbę naturalną: " << endl;
    cin >> n;
    cout << "Silnia: " << silnia_it(n) << endl;
    cout << "Silnia: " << silnia_re(n) << endl;
	return 0;
}

